scipy/stats/tests/test_stats.py

```diff
@@ -7917,6 +7917,222 @@ def test_alias(self):
+                             quantile_def[i], 
+                             confidence_two_sided[i])
+            for i in range(m)]
+        LB = [i if (not np.isnan(i)) else np.nan for i in tmp]
+        # .. UB is (N-1-LB(1-q))
+        tmp = [confint_naive(int(N[i]), 
+                             1-quantile_def[i], 
+                             confidence_two_sided[i])
+            for i in range(m)]
+        UB = [int(N[i])-1-tmp[i] if (not np.isnan(tmp[i])) else np.nan
+            for i in range(m)]
+        LB_val = [ x[i][LB[i]] if (not np.isnan(LB[i])) else np.nan 
+                for i in range(m)]
+        UB_val = [ x[i][UB[i]] if (not np.isnan(UB[i])) else np.nan 
+                for i in range(m)]
+        test_val = [(LB_val[i], UB_val[i]) for i in range(m)]
+        # .. Compare to the public function
+        tmp = [
+            stats.quantile_test(x[i], 
+                                0, 
+                                quantile_def[i], 
+                                alternative='two-sided'
+                                ).confidence_interval(confidence[i])
+            for i in range(m)]
+        public = [(tmp[i].low, tmp[i].high) for i in range(m)]
+        assert_equal(public, test_val)
+
+    def test_R_ci_quantile(self):
+        # Using the same test cases as the ci_quantile R function
+        # https://github.com/mayer79/confintr/blob/main/tests/testthat/test-ci_measures_of_location.R#L95
+
+        x = np.exp(np.arange(0, 1.01, 0.01))
+
+        lb, ub = stats.quantile_test(x, 
+                                     0, 
+                                     0.4, 
+                                     alternative='two-sided'
+                                     ).confidence_interval(0.95)
+        expected = [1.349859, 1.648721]
+        assert_almost_equal(expected, [lb,ub], decimal=5)
+
+        lb, ub = stats.quantile_test(x, 
+                                     0, 
+                                     0.5, 
+                                     alternative='two-sided'
+                                     ).confidence_interval(0.9)
+        expected = [1.506818, 1.803988]
+        assert_almost_equal(expected, [lb,ub], decimal=5)
+
+    def test_pval_ci_match(self):
+        # Verify that the following statement hold:
+        #
+        # The 95% confidence interval corresponding with alternative='less' 
+        # has -inf as its lower bound, and if xu is the upper bound, then xu 
+        # is the smallest element from the sample x such that:
+        # `stats.quantile_test(x, q=xu, p=p, alternative='less').pvalue``
+        # will be less than 5%.
+        # 
+        # And the corresponding statement for the alternative='greater' case.
+
+        # This test may fail if we the CI bound falls on a value present 
+        # multiple times in the sample. We make sure this does not happen by 
+        # generating samples via permutations rather than pure random.
+
```

Comment:
```suggestion
```
I don't see evidence of this in the test. 
In any case, why not just use a sample from a continuous distribution?

It sounds like you are saying that the test is also valid for data drawn from a discrete distribution, and repeated observations is an edge case. Please be sure to test edge cases like this against the R implementation.

Reply:
Yes, I see your point. Here is what I meant when writing that:

The root of the problem is that the CI returns the _value_ of the CI bounds, and not the corresponding index for the bound in the sample. So, if you have multiple times the same observation, and the CI bound falls on one of them, the CI will return only the value. The returned value _is_ correct, but we lose the information of the corresponding index.

The problem is that, the pvalue (like the CI) depends only on the sample indexes, not on their numerical values.  And so, because of the duplicate observation, we may be computing the pvalue fo `q=xi[i-1]` where in fact the CI boun is `xi[i]`.

(I am sorry, what I'm writing here is probably very confusing...)

In short: I'm convinced the test is correct. The fact that it may fail on repeated observation is an artifact of the implementation of the test. It can be fixed but the resulting code would be a lot messier. Using a permutation was a cheap workaround. Sampling from a continuous distribution would be another (and probably simpler) one.

Reply:
> I'm convinced the test is correct.

Here, I wasn't saying that it wasn't. I just didn't understand the code comment.

> The fact that it may fail on repeated observation is an artifact of the implementation of the test.

That's fine. Part of my point so we would also need a test showing that it does work for repeated observations.

> Sampling from a continuous distribution would be another (and probably simpler) one.

Yes. That's what I suggest in my PR to your branch.

Reply:
Switching to a continuous distribution is trivial. Making a test that also works for repeated observations isn't so simple (well, it's not that it is hard; it is just going to be more spaghetti code). 

My question is: if you think we should do the test for repeated observation, then there is no need for sampling over a continuous distribution, no? 
I mean, I can leave both tests in; it does not matter to me, but it feels somewhat redundant.

Reply:
> I mean, I can leave both tests in; it does not matter to me, but it feels somewhat redundant.

It *would* be, if you were to write a very general version of the test with repeated observations. But...

> Making a test that also works for repeated observations isn't so simple (well, it's not that it is hard; it is just going to be more spaghetti code).

In that case, we won't write such a general, "property-based" test for repeated observations. 

`test_pval_ci_match` uses randomly generated data, and the test checks that a certain property holds, regardless of what the data is. 

This is hard to write for edge cases (e.g. with repeated observations), so we'll just spot-check a few of these cases (with particular values) against the output of reference implementation.